# COMP 4151 - Assignment 4

## Due: 10/03/2019 at 11:59PM

## Name:

**Note**: 

- All work in this assignment must be done, created, originated from each student. 
- All external assistance must be explicitly mentioned and cited. 
- Using work done by others without explicit citation is considered cheating. 
- A student will receive a zero grade on the assignment for cheating. 
- Repeated offence will lead to additional consequences.

## Introduction

This assignment consists of two parts. In part I, you will analyze a dataset that contains information about people who have and do not have diabetes. Each data point has 8 features and an outcome with value of either 1 (having diabetes) or 0 (not having diabetes).

The purpose of part I is to identify the **three most important** features to be used for predicting if a new data point is of type 1 or 0.



## Part I

Data set: https://umdrive.memphis.edu/vphan/public/4151/diabetes_train.csv

In [1]:
import pandas
df_train = pandas.read_csv('/Users/XinJu/Desktop/Fall 2019/COMP 4151/Datasets/diabetes_train.csv')
df_train.head(5)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            1       89             66             23       94  28.1   
3            0      137             40             35      168  43.1   
4            5      116             74              0        0  25.6   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.167   21        0  
3                     2.288   33        1  
4                     0.201   30        0

**Question 1**: Compare the average of each attribute for outcome 1 versus outcome 0.  Show how to do this with pandas.

In [2]:
by_outcome = df_train.groupby('Outcome') #Group training dataset by Outcome (1 or 0)

In [3]:
#Average of each attribute for outcomes 1 and 0 
by_outcome.mean()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
Outcome                                                                      
0           3.266520  109.843612      67.757709      19.460352   67.557269   
1           4.945607  140.326360      70.025105      22.615063  103.230126   

               BMI  DiabetesPedigreeFunction        Age  
Outcome                                                  
0        30.219604                  0.423264  31.090308  
1        35.035983                  0.540707  37.125523

Outcome 0:
- Avg. Pregnancies is about 3.27
- Avg. Glucose is about 109.84
- Avg. Blood Pressure is about 67.76
- Avg. Skin Thickness is about 19.46
- Avg. Insulin is about 67.56
- Avg. BMI is about 30.22
- Avg. Diabetes Pedigree Function is about 0.42
- Avg. Age is about 31.09
    
Outcome 1:
- Avg. Pregnancies is about 4.95
- Avg. Glucose is about 140.33
- Avg. Blood Pressure is about 70.03
- Avg. Skin Thickness is about 22.62
- Avg. Insulin is about 103.23
- Avg. BMI is about 35.04
- Avg. Diabetes Pedigree Function is about 0.54
- Avg. Age is about 37.13

**Question 2**: Use seaborn to show and compare the boxplot of each attribute for outcomes of type 1 versus type 0.

Note:
* You should understand what a boxplot is. 
* Use "hue" to distinguish between different outcomes.
* Based on the visualization (which use should include in this assignment), draw some conclustion as to which attributes are most important.  Explain your conclusion.

In [4]:
import seaborn
from matplotlib import pyplot

Attributes Glucose, Age, and BMI appear to be the most important attributes (most likely to significantly affect the outcome), as they appear to display the largest difference in distribution of values.

##### Pregnancies vs. Outcome

In [5]:
seaborn.boxplot(x = "Outcome", y = "Pregnancies", hue = 'Outcome', data = diabetes_train)

NameError: name 'diabetes_train' is not defined

##### Glucose vs. Outcome

In [ ]:
seaborn.boxplot(x = "Outcome", y = "Glucose", hue = 'Outcome', data = diabetes_train)

##### Blood Pressure vs. Outcome

In [ ]:
seaborn.boxplot(x = "Outcome", y = "BloodPressure", hue = 'Outcome', data = diabetes_train)

##### Skin Thickness vs. Outcome

In [ ]:
seaborn.boxplot(x = "Outcome", y = "SkinThickness", hue = 'Outcome', data = diabetes_train)

##### Insulin vs. Outcome

In [ ]:
seaborn.boxplot(x = "Outcome", y = "Insulin", hue = 'Outcome', data = diabetes_train)

##### BMI vs. Outcome

In [ ]:
seaborn.boxplot(x = "Outcome", y = "BMI", hue = 'Outcome', data = diabetes_train)

##### Diabetes Pedigree vs. Outcome

In [ ]:
seaborn.boxplot(x = "Outcome", y = "DiabetesPedigreeFunction", hue = 'Outcome', data = diabetes_train)

##### Age vs. Outcome

In [ ]:
seaborn.boxplot(x = "Outcome", y = "Age", hue = 'Outcome', data = diabetes_train)

## Part II

Datasets:
* For model bulding: https://umdrive.memphis.edu/vphan/public/4151/diabetes_train.csv
* For model evaluation: https://umdrive.memphis.edu/vphan/public/4151/diabetes_test.csv

In [6]:
df_test = pandas.read_csv('/Users/XinJu/Desktop/Fall 2019/COMP 4151/Datasets/diabetes_test.csv')
df_test.sample(5)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
25            4      142             86              0        0  44.0   
37           10      122             78             31        0  27.6   
43            5      121             72             23      112  26.2   
21            4      125             80              0        0  32.3   
18            2      101             58             17      265  24.2   

    DiabetesPedigreeFunction  Age  Outcome  
25                     0.645   22        1  
37                     0.512   45        0  
43                     0.245   30        0  
21                     0.536   27        1  
18                     0.614   23        0

**Question 3**: build a decision tree model using the diabetes_train dataset.  Draw the decision tree (use draw_tree) at various max_depth and pick one that you think is good and include it here.  You should show the code of your work.

In [8]:
import draw_tree
from sklearn.tree import DecisionTreeClassifier

In [21]:
#Define predictors (X) and response (Y)
x = df_train.drop(columns = ['Outcome']) #X
y = df_train['Outcome'] #Y

In [22]:
model_tree = DecisionTreeClassifier(max_depth = 3)
model_tree.fit(x, y)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=3,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [206]:
draw_tree.visualize_tree(model_tree, x.columns, 'diabetes_train_tree')

File saved to diabetes_train_tree.png


**Question 4**: show the predictions of your decision tree model using the features of the diabetes_test dataset.


In [23]:
model_tree.feature_importances_

array([0.        , 0.59534391, 0.        , 0.        , 0.        ,
       0.27377808, 0.        , 0.13087801])

In [208]:
x.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age'],
      dtype='object')

In [209]:
df_train.head(2)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0

In [215]:
model_tree.predict([
    #Subject corresponding to index 0: (Outcome = 1 - diabetes)
    [6, 148, 72, 35, 0, 33.6, 0.627, 50],  
    #Subject corresponding to index 1: (Outcome = 0 - no diabetes)
    [1, 85, 66, 29, 0, 26.6, 0.351, 31],
    #New sample subject profile: (Outcome = 0 - no diabetes)
    [0, 150, 50, 20, 5, 17.6, 0.500, 20] 
])

array([1, 0, 0])

**Question 5**: write a Python function that takes as input a model and returns the accuracy of that model.

Notes:
* The model is built using training data and evaluated using testing data.  Two different datasets were provided for you above.
* The accuracy is the prediction is simply the number of correct predictions divided by the number of predictions.

In [238]:
def model_accuracy(model):
    import numpy 
    # Convert data frames to array (required input for model.predict())
    test_y = df_test['Outcome'].to_numpy() #Outcomes from testing data set
    test_x = df_test.drop(columns = ['Outcome']).to_numpy() #Predictors and their values from testing data set
    # (Sources: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_numpy.html)
    
    predictions = model.predict(test_x) #Make predictions on test data
    correct_predictions = 0
    total_predictions = len(predictions)
    for i in range(len(predictions)): #Iterate through predictions
        if test_y[i] == predictions[i]:
            correct_predictions = correct_predictions + 1 #If prediction is correct, add 1 to number of correct predictions
        
    accuracy = correct_predictions/total_predictions #Calculate accuracy
    return accuracy

In [239]:
model_accuracy(model_tree)

0.76

Decision Tree Model (max_depth = 3) accuracy is 76%.

**Question 6**: compare the accuracy of two methods of prediction: decision tree and k-nearest neighbor.

In [103]:
from sklearn.neighbors import KNeighborsClassifier

In [180]:
model_KNN = KNeighborsClassifier(n_neighbors=7) #Creat k-NN model

In [174]:
model_KNN.fit(x,y) #Train k-NN model on training data 

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=7, p=2,
                     weights='uniform')

In [176]:
model_accuracy(model_KNN)  #Test predictive accuracy of k-NN model 

0.8133333333333334

k-NN Model accuracy is around 81.3%.

In this case, the k-nearest neighbor classification method (using k = 7, 
which appears to optimize the predicitive accuracy of the model) provides better results than the decision tree 
classification method. 

Predictive Accuracy comparison:
- k-Nearest Neighbors (k = 7): around 81.3%
- Decision tree (max_depth = 3): 76%